File: app/src/main/java/com/example/aaron/recipeassistant/Model/InstructionListener.java
package com.example.aaron.recipeassistant.Model;

import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;
import android.widget.Toast;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;


import edu.cmu.pocketsphinx.Assets;
import edu.cmu.pocketsphinx.Hypothesis;
import edu.cmu.pocketsphinx.RecognitionListener;
import edu.cmu.pocketsphinx.SpeechRecognizer;
import edu.cmu.pocketsphinx.SpeechRecognizerSetup;

public class InstructionListener implements RecognitionListener {

    private static final String INSTRUCTION_SEARCH = "instruction";

    private static final String INGREDIENTS = "INGREDIENTS";
    private static final String PREV_DIRECTION = "PREVIOUS DIRECTION";
    private static final String NEXT_DIRECTION = "NEXT DIRECTION";
    private static final String FIRST_DIRECTION = "FIRST DIRECTION";
    private static final String FINAL_DIRECTION = "FINAL DIRECTION";

    private static final List<String> INSTRUCTIONS_LIST = Arrays.asList(INGREDIENTS, PREV_DIRECTION, NEXT_DIRECTION,
            FIRST_DIRECTION, FINAL_DIRECTION);


    private SpeechRecognizer recognizer;
    private Context context;
    private RecipeReader recipeReader;

    public InstructionListener(Context context, RecipeReader recipeReader) {
        this.recipeReader = recipeReader;
        this.context = context;
        initRecognizer();
    }

    private void initRecognizer() {
        new AsyncTask<Void, Void, Exception>() {
            @Override
            protected Exception doInBackground(Void... params) {
                try {
                    Assets assets = new Assets(context);
                    File assetDir = assets.syncAssets();
                    setupRecognizer(assetDir);
                } catch (IOException e) {
                    return e;
                }

                return null;
            }

            @Override
            protected void onPostExecute(Exception e) {
                if (e != null) {
                    Toast.makeText(context, "Failed to init recognizer " + e, Toast.LENGTH_LONG).show();
                } else {
                    Toast.makeText(context, "Recognizer initialized", Toast.LENGTH_LONG).show();

                }
            }
        }.execute();
    }

    private void setupRecognizer(File assetsDir) throws IOException {
        recognizer = SpeechRecognizerSetup.defaultSetup()
                .setAcousticModel(new File(assetsDir, "en-us-ptm"))
                .setDictionary(new File(assetsDir, "5660.dic"))
                .getRecognizer();

        recognizer.addListener(this);
        //File languageModel = new File(assetsDir, "5660.lm");
        //recognizer.addNgramSearch(INSTRUCTION_SEARCH, languageModel);
        File instructionsGrammer = new File(assetsDir, "5660.gram");
        recognizer.addKeywordSearch(INSTRUCTION_SEARCH, instructionsGrammer);

        recognizer.startListening(INSTRUCTION_SEARCH);
    }

    public void destroy() {
        if (recognizer != null) {
            recognizer.cancel();
            recognizer.shutdown();
        }
    }

    @Override
    public void onBeginningOfSpeech() {
        Log.i("stt", "beginning of speech");

    }

    @Override
    public void onEndOfSpeech() {
        //recognizer.stop();
    }

    @Override
    public void onPartialResult(Hypothesis hypothesis) {

        Log.i("stt", "partial result");
        if (hypothesis != null) {
            recognizer.stop();
            String text = hypothesis.getHypstr();
            Log.i("stt", "onPartialResult: " + text);
            recipeReader.stopReading();
            processInstruction(text);
            recognizer.startListening(INSTRUCTION_SEARCH);
        }
    }

    private void processInstruction(String instruction) {
        switch (instruction) {
            case INGREDIENTS:
                recipeReader.readIngredients();
                break;
            case NEXT_DIRECTION:
                recipeReader.readNextDirection();
                break;
            case PREV_DIRECTION:
                recipeReader.readPreviousDirection();
                break;
            case FIRST_DIRECTION:
                recipeReader.readFirstDirection();
                break;
            case FINAL_DIRECTION:
                recipeReader.readLastDirection();
                break;
        }
    }

    @Override
    public void onResult(Hypothesis hypothesis) {

    }

    @Override
    public void onError(Exception e) {
        Toast.makeText(context, e.getMessage(), Toast.LENGTH_LONG).show();
    }

    @Override
    public void onTimeout() {
    }

}

File: app/src/main/java/com/example/aaron/recipeassistant/common/voicerecognitionservice/InstructionListener.kt
package com.example.aaron.recipeassistant.common.voicerecognitionservice

import android.app.Application
import android.util.Log
import edu.cmu.pocketsphinx.*
import java.io.File
import java.io.IOException
import kotlin.concurrent.thread

class InstructionListener(private val applicationContext: Application) : RecognitionListener {

    private var recognizer: SpeechRecognizer? = null
    private var callback: (Instruction) -> Unit = {}

    private var listening: Boolean = false

    fun isListening() = listening

    fun listen(listenerCallback: (Instruction) -> Unit) {
        listening = true
        callback = listenerCallback
        initRecognizer()
    }

    fun stopListening() {
        listening = false
        detachActivity()
    }

    fun detachActivity() {
        recognizer?.cancel()
        callback = {}
    }

    private fun initRecognizer() {
        if (recognizer == null) {
            thread {
                try {
                    val assets = Assets(applicationContext)
                    val assetDir = assets.syncAssets()
                    setupRecognizer(assetDir)
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        } else {
            recognizer?.startListening(INSTRUCTION_SEARCH)
        }
    }

    private fun setupRecognizer(assetsDir: File) {
        recognizer = SpeechRecognizerSetup.defaultSetup()
                .setAcousticModel(File(assetsDir, "en-us-ptm"))
                .setDictionary(File(assetsDir, "9711.dict"))
                .recognizer

        recognizer?.addListener(this)
        val instructionsGrammar = File(assetsDir, "keyphrase.list")
        recognizer?.addKeywordSearch(INSTRUCTION_SEARCH, instructionsGrammar)
        recognizer?.startListening(INSTRUCTION_SEARCH)
    }

    override fun onBeginningOfSpeech() {
        Log.i("stt", "beginning of speech")
    }

    override fun onPartialResult(hypothesis: Hypothesis?) {
        if (hypothesis?.hypstr?.contains(STOP) == true) {
            recognizer?.cancel()
            callback(Stop())
            recognizer?.startListening(INSTRUCTION_SEARCH)
        }
    }

    override fun onEndOfSpeech() {
        recognizer?.stop()
    }

    override fun onResult(hypothesis: Hypothesis?) {
        hypothesis?.hypstr?.let {
            Log.i("stt", "onResult: $it")
            processInstruction(it)
        }
        recognizer?.startListening(INSTRUCTION_SEARCH)
    }

    override fun onError(e: Exception) {}

    override fun onTimeout() {}

    private fun processInstruction(instruction: String) {
        Log.d("stt", "Instruction: ${instruction.trim()}")
        when (instruction.trim()) {
            PLAY_INGREDIENT -> callback(PlayIngredient())
            PREV_INGREDIENT -> callback(PrevIngredient())
            NEXT_INGREDIENT -> callback(NextIngredient())
            NEXT_DIRECTION -> callback(NextDirection())
            PREV_DIRECTION -> callback(PrevDirection())
            PLAY_DIRECTION -> callback(PlayDirection())
            STOP -> callback(Stop())
            FIRST_DIRECTION -> callback(FirstDirection())
            FINAL_DIRECTION -> callback(FinalDirection())
        }
    }

    companion object {
        private const val INSTRUCTION_SEARCH = "instruction"
        private const val PLAY_INGREDIENT = "PLAY_INGREDIENT"
        private const val PREV_INGREDIENT = "PREVIOUS_INGREDIENT"
        private const val NEXT_INGREDIENT = "NEXT_INGREDIENT"
        private const val PLAY_DIRECTION = "PLAY_DIRECTION"
        private const val PREV_DIRECTION = "PREVIOUS_DIRECTION"
        private const val NEXT_DIRECTION = "NEXT_DIRECTION"
        private const val STOP = "STOP"
        private const val FIRST_DIRECTION = "FIRST_DIRECTION"
        private const val FINAL_DIRECTION = "FINAL_DIRECTION"
    }
}

sealed class Instruction
class PrevIngredient : Instruction()
class NextIngredient : Instruction()
class PlayIngredient : Instruction()
class PrevDirection : Instruction()
class NextDirection : Instruction()
class FirstDirection : Instruction()
class FinalDirection : Instruction()
class PlayDirection : Instruction()
class Stop : Instruction()
